applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/formfinding_strategy.hpp

```diff
@@ -0,0 +1,265 @@
+                    mProjectionSettings(ProjectionSetting),
+                    mrFormFindingModelPart(rFormFindingModelPart),
+                    mPrintingFormat(PrintingFormat),
+                    mWriteNewReferenceFile(WriteNewReferenceFile)
+                 {
+                    InitializeIterationIO();
+                 }
+
+            // Destructor
+            ~FormfindingStrategy() = default;
+
+        private:
+            bool SolveSolutionStep() override
+            {
+                BaseType::SolveSolutionStep();
+                if (mPrintingFormat=="all" || mPrintingFormat=="gid") mpIterationIO->FinalizeResults();
+                return true;
+            }
+
+            void UpdateDatabase(
+                    TSystemMatrixType& A,
+                    TSystemVectorType& Dx,
+                    TSystemVectorType& b,
+                    const bool MoveMesh
+                ) override
+                {
+                    BaseType::UpdateDatabase(A,Dx, b, MoveMesh);
+                    for(auto& r_node : mrFormFindingModelPart.Nodes()){
+                        // Updating reference
+                        const array_1d<double, 3>& disp = r_node.FastGetSolutionStepValue(DISPLACEMENT);
+                        array_1d<double, 3>& disp_non_historical = r_node.GetValue(DISPLACEMENT);
+
+                        disp_non_historical = disp_non_historical + disp;
+                        r_node.GetInitialPosition() += disp;
+
+                        r_node.FastGetSolutionStepValue(DISPLACEMENT) = ZeroVector(3);
+                    }
+                    ProjectVectorOnSurfaceUtility::Execute(mrFormFindingModelPart, mProjectionSettings);
+
+                    PrintResults();
+                }
+
+                FormfindingStrategy(const FormfindingStrategy& Other) {};
+
+                void EchoInfo(const unsigned int IterationNumber) override
+                {
+                    BaseType::EchoInfo(IterationNumber);
+                    mIterationNumber = IterationNumber;
+                }
+
+                void PrintResults()
+                {
+                    if (mPrintingFormat=="all"){
+                        PrintVtkFiles(mIterationNumber);
+                        PrintGiDFiles(mIterationNumber);
+                    }
+                    else if (mPrintingFormat=="vtk") PrintVtkFiles(mIterationNumber);
+                    else if (mPrintingFormat=="gid") PrintGiDFiles(mIterationNumber);
+                    else;
+                }
+
+                void FinalizeSolutionStep() override
+                {
+                    BaseType::FinalizeSolutionStep();
+                    if (mWriteNewReferenceFile) WriteNewMdpaFile();
+                }
+
+                void WriteNewMdpaFile()
+                {
+                    Matrix output_matrix;
+                    const ProcessInfo temp_process_info;
+                    for(auto& r_element : mrFormFindingModelPart.Elements()){
```

Comment:
don't you have to clear the data of the elements? Otherwise they will also be in the new mdpa

also please add a comment why this is necessary